numpy/ma/core.py

```diff
@@ -6184,17 +6184,38 @@ def isMaskedArray(x):
 
 
 class MaskedConstant(MaskedArray):
-    # We define the masked singleton as a float for higher precedence.
-    # Note that it can be tricky sometimes w/ type comparison
-    _data = data = np.array(0.)
-    _mask = mask = np.array(True)
-    _baseclass = ndarray
+
+    __singleton = None
 
     def __new__(self):
-        return self._data.view(self)
+        if self.__singleton is None:
+            # We define the masked singleton as a float for higher precedence.
+            # Note that it can be tricky sometimes w/ type comparison
+            data = np.array(0.)
+            mask = np.array(True)
+
+            # prevent any modifications
+            data.flags.writeable = False
+            mask.flags.writeable = False
+
+            masked = MaskedArray(data, mask=mask)
+            self.__singleton = masked.view(self)
+
+        return self.__singleton
 
     def __array_finalize__(self, obj):
-        return
+        if self.__singleton is None:
```

Comment:
Wouldn't it make more sense to choose path based on `obj`? If one uses `super` in `__new__` as suggested above, then the relevant case that is OK is `obj is None` - all other cases imply view casting or slicing, neither of which I think should be allowed (except if `obj is masked`) - could they just raise?